### DISCUSSION TITLE
Xavier NX USB 3.0 devices are recognized as USB 2.0

### DISCUSSION THREAD

**Discussion by hannes09** [OE4T/meta-tegra#992] [Q&A]
Hi,

we have a custom Xavier NX board and some problemes with a USB 3.0 port.

Devices plugged in into that port are recognized as USB  2.0 devices only.
The port itself seems to be okay and lsusb reports  10000M for this port.

Output form `lsusb -t `

/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=tegra-xusb/4p, 10000M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=tegra-xusb/4p, 480M
    |__ Port 3: Dev 2, If 0, Class=Hub, Driver=hub/7p, 480M
        |__ Port 1: Dev 4, If 2, Class=Miscellaneous Device, Driver=, 480M
        |__ Port 1: Dev 4, If 0, Class=Miscellaneous Device, Driver=, 480M
        |__ Port 1: Dev 4, If 1, Class=Miscellaneous Device, Driver=, 480M
        |__ Port 5: Dev 3, If 0, Class=Communications, Driver=cdc_ether, 480M
        |__ Port 5: Dev 3, If 1, Class=CDC Data, Driver=cdc_ether, 480M
`

The device on port 1 is reported with 480M, but should be USB 3.0 super speed.
With the default  Nvidia Jetpack 4.6 we dont have the problem, only with the yocto build.
[dmesg.log](https://github.com/OE4T/meta-tegra/files/8815416/dmesg.log)

**Comment by madisongh** [comment]
> With the default Nvidia Jetpack 4.6 we dont have the problem, only with the yocto build.

Is that JetPack 4.6 running on your custom carrier, or on a development kit?  If it's on your custom carrier, have you checked that your kernel config, device tree, ODMDATA setting, and pinmux configuration are the same for both the JetPack builds and your Yocto builds? Are the USB ports on your custom carrier board wired up exactly the same way on the reference carrier board on the dev kit?

The only thing I'm seeing in the kernel log you attached is a lot of messages about USB resets, like this one:
```
[    5.813430] usb 1-3: reset high-speed USB device number 2 using tegra-xusb
```
which don't look normal.  Cranking up the debug level in the relevant drivers might help in figuring out what's going on.

**Reply by hannes09** [reply]
I am also in touch with the manufacture of the custom carrier board.
What I have checked so far:

-  ODMDATA is 0xB8190000 (Default Jetson NX with emmc)
-  pinmuxing file  was provided by the manufacture of the custom carrier board and is used  without any changes,  is working with Jetpack 4.6
-  kernel config is used from the  OE4T kernel repo without any changes 
-  the exact same device tree file is working with Jetpack 4.6, the device tree file was provided by the manufacture of the custom carrier board

I am not sure if the custom carrier board is wired up exactly the same way on reference the carrier board, there might be some changes.
e. g the custom carrier has only one physical USB 3.1 port.

**Reply by hannes09** [reply]
I found out that `lshw ` shows that one of the PCI bridges has been configured with `driver=nvm`e with meta-tegra.  `lshw `  with jetpack does not show the `driver=nvme` . So I am wondering if this configuration could be the root cause for our problem?  Or is this unrelated to this kind of problem?

![f6f3a8b6-d515-4768-87e2-9ddf2139e4ab](https://user-images.githubusercontent.com/57729935/173117357-cbb249bf-ca2b-4b1b-b004-941f4e28c597.jpg)

[lshw_jetpack.log](https://github.com/OE4T/meta-tegra/files/8880786/lshw_jetpack.log)
[lshw_yocto.log](https://github.com/OE4T/meta-tegra/files/8880787/lshw_yocto.log)

**Reply by madisongh** [reply]
I don't think so, but I'm not exactly sure where `lshw` is getting its information from.  In comparing the two logs, I see the Sandisk device attached on the JetPack setup, but nothing for the Yocto setup.  Are you sure this is a USB 2 vs 3 problem, and not just some quirk of this specific device that needs a workaround?  Maybe there's some userland magic happening in Ubuntu via udev rules or something?

**Comment by hannes09** [comment]
Im pretty sure, that it is not device related. I have tested several usb 3.0 drives,  usb 3.0 hubs and cameras.  On the yocto build every  usb 3.0 device is attached to the usb 2.0 bus (every device is detected as a high speed device) But the usb 3.0 bus is presented, firmware is okay, device tree file is okay. I'm now starting to debug inside the kernel.  I will also check for special udev  rules.
